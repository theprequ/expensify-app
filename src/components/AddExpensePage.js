import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startAddExpense } from "../actions/expenses";

export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.startAddExpense(expense);
        this.props.history.push("/");
    };
    render() {
        return (
            <div>
                <h1>Add Expense</h1>
                <ExpenseForm onSubmit={this.onSubmit} />
            </div>  
        );
    };
}

const mapDispatchToProps = (dispatch) => ({
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
});

// As of S12 L124
// Passing undefined instead of just leaving it empty
export default connect(undefined, mapDispatchToProps)(AddExpensePage);

// onSubmit is using reverse data flow (Section 4, lecture 37)
// By passing a method WITH argument from parent to child, child component can assing value to that method (in this case an expense object)
// Parent component can access those values by using the method's argument (in this case, argument is expense)