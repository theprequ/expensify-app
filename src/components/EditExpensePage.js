import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense, startRemoveExpense } from "../actions/expenses";

// This component was changed to class based component in:
// Section 12, Lecture 125
export class EditExpensePage extends React.Component {
    // (expense) comes from ExpenseForm by using reverse data flow
    onSubmit = (expense) => {
        // inside startEditExpense: this.props.expense.id is an actual prop key
        // inside startEditExpense: expense comes from ExpenseForm
        this.props.startEditExpense(this.props.expense.id, expense);
        this.props.history.push("/");
    };
    onRemove = () => {
        this.props.startRemoveExpense({ id: this.props.expense.id });
        this.props.history.push("/");
    };
    render() {
        return (
            <div>
                <ExpenseForm 
                    expense={this.props.expense}
                    onSubmit={this.onSubmit} 
                />
                <button onClick={this.onRemove}>Remove</button>
            </div>
        );
    };
}

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
});

// Below code "adds" startEditExpense and startRemoveExpense to props
// Now, "this.props" will include startEditExpense and startRemoveExpense actions and
// bunch of other methods related to normal props
// http://prntscr.com/lfnq9e
const mapDispatchToProps = (dispatch, props) => ({
    startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
    startRemoveExpense: (data) => dispatch(startRemoveExpense(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
