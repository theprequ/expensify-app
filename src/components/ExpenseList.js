import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

// Normal stataless component
// This is exported only for testing purposes
export const ExpenseList = (props) => (
    <div>
        {
            props.expenses.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.expenses.map((expense) => {
                    return <ExpenseListItem key={expense.id} {...expense} />
                })
            )
        }
    </div>
);

// Normal JavaScript function
// Here we define the things we want to get from the store
const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

// here (ExpenseList) is the component where we want the state to be shown
export default connect(mapStateToProps)(ExpenseList);