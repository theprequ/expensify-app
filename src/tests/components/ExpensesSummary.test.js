import React from "react";
import { shallow } from "enzyme";
import {ExpensesSummary} from "../../components/ExpensesSummary";

test("should correctly render ExpensesSummary with one expense", () => {
    const wrapper = shallow(<ExpensesSummary 
        expenseCount={1} 
        expensesTotal={5000} 
        />);

    expect(wrapper).toMatchSnapshot();
});

test("should correctly render ExpensesSummary with multiple expenses", () => {
    const wrapper = shallow(<ExpensesSummary 
        expenseCount={5} 
        expensesTotal={10000} 
        />);

    expect(wrapper).toMatchSnapshot();
});