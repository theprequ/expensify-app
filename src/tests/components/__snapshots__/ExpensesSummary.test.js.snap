// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render ExpensesSummary with multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    5
     
    expenses
     totalling 
    $100.00
     
  </h1>
</div>
`;

exports[`should correctly render ExpensesSummary with one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $50.00
     
  </h1>
</div>
`;
